### DISCUSSION TITLE
Pico Micropython Wake early from machine.lightsleep

### DISCUSSION THREAD

**Discussion by pollardd** [micropython/micropython#11089] [RP2040 / Pico] [answered]
I’m trying to figure out how to interrupt a machine.lightsleep on a GPIO event.
That is I’d like the pico to sleep to save power for 45minutes unless a button press is received from pin 19.
Note: light sleep disconnects thonny even after it wakes up again.  This is why I log to a file and flash the Led.
Any ideas how I can end the lightsleep prematurely by pressing a GPIO connected button?
See
#wake here if button is pressed 

```
import machine
import time
import debug
 
#led = machine.Pin("LED", machine.Pin.OUT) # Wifi Board
led = machine.Pin(25, machine.Pin.OUT) # Non Wifi Board
 
DEBUG=1
LOGTOFILE = True
 
# Pin Definitions
gateSwitch =  Pin(14, Pin.IN, Pin.PULL_UP)  # Physical Pin 19 Gnd = 18
 
 
def flash(number):
    count=1
    while(count <= number):
        led.on()
        print("LED ON")
        time.sleep(0.5)
        led.off()
        print("LED OFF")
        time.sleep(0.5)
        count = count+1
    time.sleep(2)
 
def getGateSwitch():
    # Invert the 1-0 so that 1=button pressed 0=not pressed
    return not gateSwitch.value()
 
while True:
    flash(2)
    if(DEBUG >=1):
        debug.debug(DEBUG, "main()    ", "Light Sleep Start", LOGTOFILE)
 
    machine.lightsleep(2700000)  # Sleep 45 Minutes
   
    # Wake here if switch is pressed
 
    if(DEBUG >=1):
        debug.debug(DEBUG, "main()    ", "Light Sleep Complete", LOGTOFILE)
       
    if gateSwitch()==1:
            if(DEBUG >=1):
                debug.debug(DEBUG, "main()    ", "Gate is Open", LOGTOFILE)
```

**Comment by peterhinch** [comment, marked as answer]
I think you'll need to wake up, say once per second, and check the pin.

The additional power consumption is probably quite small: if checking a pin takes 20μs and uses 20mA then the current consumption is `20*20/1_000_000`mA i.e. 0.4μA. Even if my guesses are out by a couple of orders of magnitude, the penalty is far less than the light sleep current.

**Reply by pollardd** [reply]
The example I have shown is just something I came up with as a test.  I know how unpopular questions with no examples are :)  My actual code was a bit hard to cut down.
My current work in progress program wakes up every 10 seconds to check the pin and I was looking for a better way.  Based on a post from a couple of years ago (with a very familiar face :) I thought there may be a way to do it now.  That thread also ended with basically "It's in development"
https://forum.micropython.org/viewtopic.php?f=21&t=9990

**The Project**
The pico will be hooked to a switch on my front gate that is 300m from the house.
I have a LoRa module communicating with another Pico in my house to play a tone and record information when the gate is opened.  The 10 seconds is how quick someone could get a car through the gate if someone else opens it.  It's not meant to catch criminals but let me know if someone is coming or entered while I'm out.
This is all working so far I just need to extend the battery life.
Initially the battery was only lasting about 6 hours but setting the LoRa module to standby and changing from time.sleep to machine.lightsleep has made great improvements.  Current test is up to 24 hours and its still going.  Next I'll need to stick it out in the sun and see if the little solar panel I have can keep up on a mixture of cloudy and sunny days.

Thanks for your response.  I'll post again here if I come up with something or if the current combination is good enough.
David

**Reply by peterhinch** [reply]
That forum post was in the very early days of the Pico. As you've found, `lightsleep` now works.

**Reply by pollardd** [reply]
The above change to lightsleep() and a reduction in the amount of logging increased the run time from about 6 hours to 71 hours so I’m very happy with that. :)
Hopefully my little solar panel can suck up enough power during daylight hours even if there is no direct sun light.
If not then I’ll get a bigger battery.  The current battery is 600 mAh (3.7v)
